### PR TITLE
feature: context menu

### PR DESCRIPTION
**Pull request by Jocs**
| Q                | A
| ---------------- | ---
| Bug fix?         | no
| New feature?     | yes
| BC breaks?       | no
| Deprecations?    | no
| New tests added? | not needed
| Fixed tickets    | #169 
| License          | MIT

### Description

[Description of the bug or feature]

* [x] copy \ cut \ paste \ copyAsMarkdown \ copyAsHtml \ pasteAsPlainText
* [x] insert Paragraph before or after
* [x] insert colums and rows in table block

--

### CODE REVIEW

File: src/editor/contentState/paragraphCtrl.js [hunk 1]
```diff
@@ -481,6 +481,32 @@ const paragraphCtrl = ContentState => {
     const selectionChanges = this.selectionChange()
     this.eventCenter.dispatch('selectionChange', selectionChanges)
   }
+
+  ContentState.prototype.insertParagraph = function (location) {
+    const { start, end } = this.cursor
+    // if cursor is not in one line or paragraph, can not insert paragraph
+    if (start.key !== end.key) return
+    let block = this.getBlock(start.key)
+    console.log(block)
```
Comment (fxha) [inline src/editor/contentState/paragraphCtrl.js]: .

File: src/editor/contentState/paragraphCtrl.js [hunk 2]
```diff
@@ -481,6 +481,32 @@ const paragraphCtrl = ContentState => {
...
+    if (block.type === 'span') {
+      block = this.getParent(block)
+    } else if (/th|td/.test(block.type)) {
+      block = this.getParent(this.getParent(this.getParent(this.getParent(block))))
```
Comment (fxha) [inline src/editor/contentState/paragraphCtrl.js]: I would add a comment that you want get the table.

File: src/editor/contentState/tableBlockCtrl.js [hunk 3]
```diff
@@ -202,6 +203,149 @@ const tableBlockCtrl = ContentState => {
     }
   }
 
+  // insert/remove row/column
+  ContentState.prototype.editTable = function ({ location, action, target }) {
+    const { start, end } = this.cursor
+    const block = this.getBlock(start.key)
+    if (start.key !== end.key || !/th|td/.test(block.type)) {
+      throw new Error('Cursor is not in table block, So you can not insert/edit row/column')
```
Comment (fxha) [inline src/editor/contentState/tableBlockCtrl.js]: I would guess `...block, so you...` or `...blocks. So you...` is correct.

File: src/renderer/contextMenu/menuItems.js [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+  label: 'Paste As Plain Text',
+  id: 'pasteAsPlainTextMenuItem',
+  click (menuItem, browserWindow) {
+    contextMenu.pasteAsPlainText()
+  }
+}
+
+export const INSERT_BEFORE = {
+  label: 'Insert Paragraph Before',
+  id: 'insertParagraphBefore',
```
Comment (fxha) [inline src/renderer/contextMenu/menuItems.js]: inconsistent id names

File: src/editor/contentState/tableBlockCtrl.js (hunk 3 above)
Comment (Jocs) [reply src/editor/contentState/tableBlockCtrl.js]: @fxha What's the meaning of `...block, so you... or ...blocks. So you...`?

Comment (Jocs) [reply src/editor/contentState/tableBlockCtrl.js]: 👍

**Comment by Jocs** [comment]
> Past as Plain Text does not work; it is inserted as markdown

Paste as Plain Text would paste the `markdown content`, and it will be previewed. What do you think the correct behavior of `paste as plain text`?

>removing table header will throw exception(s) and do not remove row

fixed

>removing last (right) cell will throw exception

We can only remove row or column, so how to remove last cell?

**Comment by fxha** [comment]
> Paste as Plain Text would paste the markdown content, and it will be previewed. What do you think the correct behavior of paste as plain text?

Yes, you are right.

> We can only remove row or column, so how to remove last cell? Do you mean that when the column number is less than 2(or equal), we can not remove any columns?

Sorry my fault. column not cell.

Another problem:

`Ctrl+Shift+V` paste content as not-formatted text into Mark Text and `Ctrl+V` as formatted content  (like in v.0.10.21). But the last (formatted past) does not work any more. E.g. copy Wikipedia article with headings, links etc and past it into v.0.10.21 and master. You will notice the differences.

**Comment by Jocs** [comment]
>Ctrl+Shift+V paste content as not-formatted text into Mark Text and Ctrl+V as formatted content (like in v.0.10.21). But the last (formatted past) does not work any more. E.g. copy Wikipedia article with headings, links etc and past it into v.0.10.21 and master. You will notice the differences.

Yes it's a issue that I introduced when I add new features.

**Pull request merged**